swarm/chunk/tags.go

```diff
@@ -0,0 +1,233 @@
+)
+
+// Tag represents info on the status of new chunks
+type Tag struct {
+	uid       uint32     // a unique identifier for this tag
+	name      string     // a name tag for this tag
+	total     uint32     // total chunks belonging to a tag
+	split     uint32     // number of chunks already processed by splitter for hashing
+	stored    uint32     // number of chunks already stored locally
+	sent      uint32     // number of chunks sent for push syncing
+	synced    uint32     // number of chunks synced with proof
+	startedAt time.Time  // tag started to calculate ETA
+	State     chan State // channel to signal completion
+}
+
+// New creates a new tag, stores it by the name and returns it
+// it returns an error if the tag with this name already exists
+func NewTag(uid uint32, s string, total uint32) *Tag {
+	if len(s) == 0 {
+		s = fmt.Sprintf("upload_%d", time.Now().Unix())
+	}
+	t := &Tag{
+		uid:       uid,
+		name:      s,
+		startedAt: time.Now(),
+		total:     uint32(total),
+		State:     make(chan State, 5),
+	}
+	return t
+}
+
+// Inc increments the count for a state
+func (t *Tag) Inc(state State) {
+	var v *uint32
+	switch state {
+	case SPLIT:
+		v = &t.split
+	case STORED:
+		v = &t.stored
+	case SENT:
+		v = &t.sent
+	case SYNCED:
+		v = &t.synced
+	}
+	n := atomic.AddUint32(v, 1)
+	if int(n) == t.GetTotal() {
+		t.State <- state
+	}
+}
+
+// Get returns the count for a state on a tag
+func (t *Tag) Get(state State) int {
+	var v *uint32
+	switch state {
+	case SPLIT:
+		v = &t.split
+	case STORED:
+		v = &t.stored
+	case SENT:
+		v = &t.sent
+	case SYNCED:
+		v = &t.synced
+	}
+	return int(atomic.LoadUint32(v))
+}
+
+// GetUid returns the unique identifier
+func (t Tag) GetUid() uint32 {
+	return t.uid
+}
+
+func (t Tag) GetName() string {
+	return t.name
+}
+
+// GetTotal returns the total count
+func (t *Tag) GetTotal() int {
+	return int(atomic.LoadUint32(&t.total))
+}
+
+// SetTotal sets total count to SPLIT count
+// is meant to be called when splitter finishes for input streams of unknown size
+func (t *Tag) SetTotal() int {
+	total := atomic.LoadUint32(&t.split)
+	atomic.StoreUint32(&t.total, total)
+	return int(total)
+}
+
+// Status returns the value of state and the total count
+func (t *Tag) Status(state State) (int, int) {
+	return t.Get(state), int(atomic.LoadUint32(&t.total))
+}
+
+// ETA returns the time of completion estimated based on time passed and rate of completion
+func (t *Tag) ETA(state State) (time.Time, error) {
+	cnt := t.Get(state)
+	total := t.GetTotal()
+	if cnt == 0 || total == 0 {
+		return time.Time{}, errNoETA
+	}
+	diff := time.Since(t.startedAt)
+	dur := time.Duration(total) * diff / time.Duration(cnt)
+	return t.startedAt.Add(dur), nil
+}
+
+// WaitTill blocks until count for the State reaches total cnt
+func (tg *Tag) WaitTill(ctx context.Context, s State) error {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case c := <-tg.State:
```

Comment:
What would happen if there are multiple WaitTill calls? Only one of them will receive a State from tg.State channel, having others not returning correctly. This would be a problem for WaitTill calls with the same State argument, but also with different State arguments but with flakiness.

Maybe this function needs tests to verify that.

Reply:
TBH i took this notification logic from @zelig's original [push sync branch](https://github.com/ethersphere/go-ethereum/pull/971/files#diff-49ddfbee8a37fd7fbd0d4b1f69425ba4). Looking at this more and more makes me understand that this notification logic is not feasible, not just because of the reason you made here, but also because that once a value that's passed on a channel does not match the value a caller is waiting for - the value is just forgotten and other callers that might wait on this value will just deadlock on this.
Example:
1. `goroutine a` calls `WaitTill(ctx, 0)`
2. `goroutine b` calls `WaitTill(ctx, 1)`
3. `goroutine b` gets scheduled before `goroutine a`
4. `...` (chunker is chunking the tag)
5. value `0` is passed on the state channel and gets picked up by `goroutine b`. does not match `if` statement
6. value `1` either:
6.1. gets picked up by `goroutine a`. does not pass `if` statement. both goroutines block till timeout
6.2. gets picked up by `goroutine b`. passes `if` statement and returns `nil`. `goroutine a` blocks till timeout.

I think this whole notification logic is convoluted and can be removed at the time being. We should consider and implement a notification logic that relies on polling instead, and should propagate from the HTTP package (although the idea of having this on such low level is very romantic).
Another approach would be to implement with a stack of channels to which notification should be made (fan-out pattern), or, alternatively, maintaining a bunch of pipes for each state and close them when state is reached. I was looking into `sync.Cond` but it doesn't fit this use case.

Reply:
This should only be used in tests and assume a single caller. 
but TBH best if removed and implement polling `Status` and compare if equal either in tests if needed or here